### ISSUE TITLE
Database sync should start automatically when peers connect

### ISSUE THREAD

**Issue by beingmohit** [orbitdb/orbit-db#237]
I am running 2 nodes on two different machines. First node starts and adds a entry in eventlog and keeps on running. Then Second node starts and fetch all eventlog entries and prints every 3 seconds.

In this scenarios, Second node only prints empty arrays. 

But If I change code of first node to add a entry every 3 seconds. Second node starts getting all entries.

So I deduced,  When a new node starts, it won't be able to get any data unless some other node pushes something to eventlog. I think, its becuase new nodes will not have any heads sync unless other nodes pushes something. 

Is my understanding right? Is there any workaround for this other then periodically pushing data to eventlog?

**Comment by beingmohit** [comment]
One solution I come up with is, when a new node starts, it will push something to eventlog, so that other listening peers can connect with it, and sync but it doesn't seems to be working.

**Comment by beingmohit** [comment]
Does orbitdb have any way to periodically push heads without saving these pubs into eventlog ? If not, is it good idea to add this feature? I will create PR.

**Comment by haadcode** [comment]
@beingmohit Thanks you for opening the issue on this! This is indeed a problem currently and what you describe is correct.

Periodically pushing updates might not be the best solution though, and we definitely don't want extra "update" entries in the log.

We previously used this https://github.com/libp2p/js-libp2p-floodsub/pull/32 to solve the problem but it looks like it won't end up in floodsub as a feature, so it has been suggested by @diasdavid that we would do that on app layer, ie. in orbit-db.

I haven't looked into what exactly needs to be done here, but from functionality perspective, we need a way to determine when "new" peers connect (in pubsub) and send them our latest local entry. 

This seems like a good place to start to determine when peers connect: https://github.com/ipfs/js-ipfs/blob/master/src/core/components/pubsub.js#L75.

Perhaps we can keep track of peers in https://github.com/orbitdb/orbit-db-pubsub/blob/master/src/ipfs-pubsub.js and use callbacks back to https://github.com/orbitdb/orbit-db/blob/master/src/OrbitDB.js and then directly send a message to the peer (instead through publish() in pubsub).

Sorry if this doesn't help much or sounds vague, I haven't had time to properly think about it yet.

@beingmohit would you be interested to think this through and dig deeper as to what needs to be done here?

(I changed the topic to reflect the issue described in OP)

**Comment by beingmohit** [comment]
@haadcode Ok, I will think this through and come up with possible solutions.

**Comment by haadcode** [comment]
@beingmohit awesome, thank you! ❤️ 

If you need any help, best to jump on [#orbitdb](https://webchat.freenode.net/?channels=%23orbitdb) in IRC or post questions here. Don't hesitate to ask! 😄

**Comment by haadcode** [comment]
@beingmohit we could probably use https://github.com/ipfs-shipyard/ipfs-pubsub-room that @pgte implemented. Wanna take a look and see if it's what's needed and how it could be integrated to https://github.com/orbitdb/orbit-db-pubsub?

**Comment by haadcode** [comment]
@beingmohit did you have a chance to take a look at this?

I started to look into it to see if we can use `pfs-pubsub-room` (I believe we can) and made a quick example: https://github.com/orbitdb/orbit-db-pubsub/commit/fdf5202e26473d5920d082cddd3a71d3c951b9e8.

There's a small bug in ipfs-pubsub-room atm which I reported [here](https://github.com/ipfs-shipyard/ipfs-pubsub-room/issues/16) and as that gets fixed, it should be all good.

In addition to the changes in orbit-db-pubsub, we need to get the callback when a new peer was connected and send them the latest heads we have. This is done in orbit-db/OrbitDB.js by adding a callback to the subscribe function in `_createStore()`, eg:
```js
this._pubsub.subscribe(addr, this._onMessage.bind(this), this._onPeerConnected.bind(this))
```
and adding that callback function to OrbitDB class:
```js
  _onPeerConnected (address, peer, room) {
    const store = this.stores[address]
    if (store) {
      let heads = store._oplog.heads
      if (heads.length > 0) {
        room.sendTo(peer, new Buffer(JSON.stringify(heads)))
      }
    }
  }
```

I think that should do it. I might be overlooking some details, but it's working on my machine.

@beingmohit would you be interested to champion this all the way to a PR across all the different repos and land it in orbit-db? You can absolutely say no, too! :) Let me know.

**Comment by beingmohit** [comment]
@haadcode I love to do this stuff. But i am unable to spare time for this. Please do you rely on me.

**Comment by haadcode** [comment]
Got it. I'll work on this over the next few days and keep the thread updated on the progress.

**Comment by haadcode** [comment]
This has now been fixed in [v0.18.0](https://github.com/orbitdb/orbit-db/issues/256)! Thank you everyone for reporting it and helping to fix it! ❤️
